### PR TITLE
smolyak index grid

### PR DESCRIPTION
**Pull request by mzf0069**
Added a method to generate the unique indexes for each level. 
Added a method to generate the index of grid points.

### CODE REVIEW

**Comment by cakieslich** [comment]
Hello world

File: smolyay/smolyak.py
Comment (mzbush) [inline smolyay/smolyak.py:30]: Technically, the index_per_level can describe more levels than is required for the exactness. If you only want to generate the index_per_level once and use that for multiple values of exactness, you can just generate it for the highest value for exactness and then never need to generate it again. This should only check if the index_per_level describes less than the required amount.

If the index_per_level cannot contain extra information, then a check could be used to take only the required portion in a new list.

File: smolyay/smolyak.py
Comment (mzbush) [inline smolyay/smolyak.py:44]: Here the code requires that index_per_level cannot contain extra information. This part might be edited to avoid this.

File: smolyay/smolyak.py
Comment (mzbush) [inline smolyay/smolyak.py:63]:
```suggestion
        First, it generates allowed compositions of levels
```

File: smolyay/smolyak.py
Comment (mzbush) [inline smolyay/smolyak.py:87]: For the possible level_compositions, is there ever going to be a reason that we want the code to print that out so it can be seen for writing a paper or any other reason? If so, this function may need to be split into multiple functions.

File: smolyay/smolyak.py
Comment (mzbush) [inline smolyay/smolyak.py:155]: This for loop is very nested. May want to avoid that for optimization purposes.

**Review by mzbush — CHANGES_REQUESTED** [review]
Most things look good, besides a few tweaks to make it better.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py]: Addressed.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py]: Addressed.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:88]: For the case of generating grid points, I don't think it's necessary; however, we can split it in case it's needed.

**Review by mzbush — APPROVED** [review]
Everything looks fine. I am happy.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:38]: I think this whole function can all be done in a more readable way like this
```py
# cumulative sum to get end index of each level, up to exactness+1
end_levels = numpy.cumsum(self.index_per_level[:self.exactness+1])
# create ranges of indexes at each level
level_indexes = [list(range(end-n, end)) for end, n in zip(end_levels, self.index_per_level)]
return level_indexes
```

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:56]: This will also become a property, so it should get a short docstring like above. The comments about NEXCOM, etc. belong up in the detailed class description, which still needs to be written.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:77]: As we discussed, we should probably pull this out into a function

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:1]: In published code, you should try not to alias packages unless you really need to because it pollutes the namespace:
```suggestion
import numpy
```
In your own scripts, it's perfectly fine to do this though

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:111]: I have a really hard time following the code from here onward, but I think that what you are doing can be achieved in a simpler way using `numpy.meshgrid`. See https://stackoverflow.com/questions/1208118/using-numpy-to-build-an-array-of-all-combinations-of-two-arrays for an explanation:

You would define `points = None` at the start of the function, then do something like this inside the for loop:
```py
            # get all combinations of points at each level
            for composition in compositions:
                point_ids = [level_indexes[i] for i in composition]
                points_ = np.array(np.meshgrid(*point_ids)).T.reshape(-1, self.dimension)
                if points is None:
                    points = points_
                else:
                    points = np.concatenate((points,points_), axis=0)
```
`points` will be a NumPy array, so you can either convert to a list using the `tolist()` method or leave as is

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:6]: Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:27]: Created a `_update` function and moved it there.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:37]: Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:38]: Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:8]: Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:77]: Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:1]: Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:111]: Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:87]: Done.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:157]: Don't commit debugging, these belong in tests though!
```suggestion
```

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:106]: Function arguments should be descriptive and make sense for the function. In this case `generate_compositions` is a function that just does NEXCOM, and it is not really tied to the `IndexGrid` class. You could use `n, k` to match the variables used in the algorithm or `value, num_parts` if you want to be more descriptive.

Further, I think you should add a third boolean argument like `weak` or `include_zero`. If `True`, this function should include zeros in the composition.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:137]: I think the `+1` is incorrect here. I would also only use one check or the other, as these are redundant. Technically, it would be faster to use only the first one, since then you don't have to precompute the number of compositions.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:151]: Rather than save this into an array, you can use Python's `yield` syntax to turn this into a generator. That way, you don't actually have to generate all the compositions in memory (if you don't want to).
```suggestion
        yield r
```
You would then delete the return statement below and the `compositions` variable above. This is convenient because you can easily do:
```
for c in generate_compositions(n,k):
    print c
```
without generating all the compositions at once, but also
```
compositions = list(generate_compositions(n,k))
```
if you do want them.

Don't forget to add 1 to everything here if you are doing strong composition (no zeros).

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:125]: These can be removed with the suggestions below

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:129]: I would suggest to use variable names consistent with the paper the algorithm came from, it is easier to read later on.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:121]: Here, you would use the third (boolean) argument to decide what the max value used by the algorithm was

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:55]: Stylistically, I like the code under a comment to all appear together
```suggestion
        # get all combinations of points at each level
        points = None
```

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:56]: You should either use a more descriptive variable name than `summ` or a traditional index variable like `i`.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:57]: Per PEP 8 (I think), don't write comments for things that are obvious
```suggestion
```

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:59]: If you use the generator style I suggested, this will work:
```suggestion
            for composition in generate_compositions(summ, self.dimension, include_zeros=False):
```

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:60]: Is `point_ids` the right name for this? I just chose something random, so you should check that. Also, be careful here because `i` is used as an iterator inside an outer loop. You might want to switch to `j` or something else like `c` to make sure it doesn't accidentally get used

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:61]: A comment would be helpful here that this is a code for generating all combinations of the arrays along each dimension

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:73]: The docstring needs to be shortened for both properties (see old suggestion). Properties generally only get one-liners.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:29]: You will want to turn each of these into a property as well, which has both a getter and a setter. The setters should make `self._needs_update = True`.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:8]: Do not include these in your files. The top part is not necessary, and we will write docstrings for the modules.

**Review by mphoward — CHANGES_REQUESTED** [review]
This is a good improvement! Here are some further comments and suggestions

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:157]: Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:106]: Done. 

Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:8]: Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:121]: Done.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:129]: In the paper, they used simply "n" instead of "max_level_index". However, since "n" differs depending on zero being included or not, I named it max_value instead.

File: smolyay/smolyak.py
Comment (mzf0069) [reply smolyay/smolyak.py:59]: Done.

File: smolyay/smolyak.py
Comment (mphoward) [reply smolyay/smolyak.py:129]: Right, but `max_level_index` is meaningless to `generate_compositions` because this function doesn't know anything about `IndexGrid`. The math problem is usually defined as "the compositions of n into k parts", which is why NEXCOM uses the names it does. I suggest the same here.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:35]: Or this:
```suggestion
        """int: Level of exactness."""
```

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:60]: You should document any exceptions that you might raise (e.g., IndexError) here in a `Raises` section.

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:79]: This makes it clearer that the range is from dimension up:
```suggestion
                                self.dimension+self.exactness+1):
```

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:78]: An iterator should be singular
```suggestion
        for sum_level in range(self.dimension,
```

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:116]:
```suggestion
    """Generate compositions of a value into parts.
```

File: smolyay/smolyak.py
Comment (mphoward) [inline smolyay/smolyak.py:139]: I think this variable is not needed?
```suggestion
```

**Review by mphoward — CHANGES_REQUESTED** [review]
Great work! Here are a few more small things, but then this mostly just needs testing.

File: tests/smolyay_test.py
Comment (mphoward) [inline tests/smolyay_test.py:6]: Rather than make this a global variable, you can make it a test fixture:
```py
@pytest.fixture
def expected_points():
    return numpy.array(...)

def test_something(expected_points):
    assert point in expected_points
```

File: tests/smolyay_test.py
Comment (mphoward) [inline tests/smolyay_test.py:55]: Can't this be done with:
```
    assert sorted(test_class.grid_point_index) == grid_points_expect
```
?

File: tests/smolyay_test.py
Comment (mphoward) [inline tests/smolyay_test.py:196]: Give this a more descriptive name: what is it testing for an error in?

**Review by mphoward — CHANGES_REQUESTED** [review]
Thank you! I have some comments on both the main code and the tests. It would be good if you both can make a plan to implement these.

**Comment by mphoward** [comment]
I will wait for @mzbush to update the tests before I review, as I think that should be close to final pass. Please let me know when you're ready for that!

File: tests/smolyay_test.py
Comment (mphoward) [inline tests/smolyay_test.py:29]: I think that `array_equiv` is a weaker test so this should be `array_equal`. Was it failing otherwise?
```suggestion
    assert numpy.array_equal(test_class.index_per_level,[1,2,3])
```
In general, these tests make clear that there is a mixture of lists and NumPy arrays being used in the class. Do we want to perhaps use only lists? That would allow these tests to simplify to using `==` operators, and I could imagine that being helpful elsewhere. It should be as easy as converting the NumPy array `tolist` before storing it in the class.

File: tests/smolyay_test.py
Comment (mphoward) [inline tests/smolyay_test.py:76]: Perform each of these assertions separately
```suggestion
    assert not numpy.array_equiv(grid_points_1,grid_points_2)
    assert numpy.array_equiv(grid_points_2, expected_points)
```

File: tests/smolyay_test.py
Comment (mphoward) [inline tests/smolyay_test.py:159]: This would be an easier way to run the generator:
```suggestion
        list(generate_compositions(6,7,include_zero=False))
```

**Review by mphoward — CHANGES_REQUESTED** [review]
A couple more small changes from my side, then I think this is ready for Dr. K to review and merge

File: tests/smolyay_test.py
Comment (mphoward) [inline tests/smolyay_test.py:34]:
```suggestion
```

File: tests/smolyay_test.py
Comment (mphoward) [inline tests/smolyay_test.py:41]:
```suggestion
```

File: tests/smolyay_test.py
Comment (mphoward) [inline tests/smolyay_test.py:48]:
```suggestion
```

File: tests/smolyay_test.py
Comment (mphoward) [inline tests/smolyay_test.py:74]:
```suggestion
    assert grid_points_1 != grid_points_2
```

**Review by mphoward — APPROVED** [review]
Great job! I have marked a couple minor final changes, which you can batch apply and commit using the GitHub web interface if you agree. Otherwise, lgtm.

@cakieslich when you review and approve, I recommend squash-merging this given the long commit history. @mzf0069 and @mzbush , if this is how this branch is merged, we may need to do a rebase of your other PRs before they can be merged. We can discuss how to do that later.

File: tests/smolyay_test.py
Comment (mzbush) [reply tests/smolyay_test.py:143]: Funny thing about this part of the code. When I was making it for some reason using the list method didn't work when include_zero was true but it did work when include_zero was false. @mzf0069 and I have no idea why this is the case. That's why composition_obtained is generated differently.

Comment (mphoward) [reply tests/smolyay_test.py:143]: Ah, that indicates that there is an issue with the code to resolve. If the test is valid (and the change I suggested is certainly the way a generator is expected to work), you should modify the code to pass the test, not hack the test until the code works. Let us know if you need help debugging.

Comment (mzbush) [reply tests/smolyay_test.py:137]: Should I update the test or wait for @mzf0069 to fix the code so it passes the test?

Comment (mphoward) [reply tests/smolyay_test.py:137]: It doesn’t matter to me! One of you could take care of both, you do not need to be separate.

**Comment by mphoward** [comment]
I believe this PR is ready to squash merge @cakieslich

**Review by cakieslich — APPROVED** [review]

**Pull request merged**
